### PR TITLE
Fix some memory leaks

### PR DESCRIPTION
**Pull request by Kanma**
Hello

While using valgrind on my own software, I got some memory leaks notifications in astrometry.net, so I took some time to fix them.

The results I get on my test data is the same, so hopefully there isn't an edge-case I didn't saw somewhere.

### CODE REVIEW

**Comment by Kanma** [comment]
Ok, I did break things...

**Comment by aMarcireau** [comment]
@Kanma I did the same thing and found a handful of other memory leaks: https://github.com/neuromorphicsystems/astrometry/blob/main/astrometry.net.patch

I can open a new PR once yours is merged, unless you would like to integrate these changes into your PR directly.

**Comment by dstndstn** [comment]
Like I said before, you are both trying to change the "&testperm" to "NULL", and that changes the behavior in a bad way.  You must change that before this can be merged.

**Comment by Kanma** [comment]
Hello

Sorry, I can't find any trace of that message, where did you said that? I will look at that as soon as I can.

File: solver/tweak2.c [hunk 1]
```diff
@@ -526,7 +528,7 @@ sip_t* tweak2(const double* fieldxy, int Nfield,
                                         W, H, distractors,
                                         logodds_bail, LARGE_VAL,
                                         &besti, &odds, &theta, NULL,
-                                        &testperm, &refperm);
+                                        NULL, &refperm);
```
Comment (dstndstn) [inline solver/tweak2.c:531]: as above, this can't be changed like this

**Comment by dstndstn** [comment]
I'm sorry!  I had, apparently, what github thought was a half-finished review!, so it showed *me* my comments, but it didn't show them to *you* or anyone else!

Thank you for this/these pull requests!

**Comment by aMarcireau** [comment]
The change to *solver.c* (`if (mo.sip != solver->best_match.sip)`) does not completely fix the SIP memory leak. I believe that the root of the issue is that SIP objects are not reallocated when match objects are copied.

https://github.com/dstndstn/astrometry.net/blob/04e97d5365525ac429dca69b344f9f4d9070b739/solver/solver.c#L1643

In other words, match objects share SIP pointers, which makes it very hard to tell which match object is responsible for freeing the shared SIP.

I think that systematically re-allocating the SIP when copying match objects is a reasonable fix (though it may incur a small performance drop since the library would allocate more SIP objects).

Under this fix, the `mo.sip` in *solver.c* would still need to be freed (but `solver->best_match.sip` would point to its own copy so there would be no need to check that they match).

Besides changing the copy behavior, the library would also need to free the SIP pointer whenever the match object is freed.

File: solver/verify.c [hunk 2]
```diff
@@ -1593,9 +1593,9 @@ double verify_star_lists(double* refxys, int NR,
         *p_besti = besti;
 
     if (p_testperm)
-        *p_testperm = v.testperm;
-    else
-        free(v.testperm);
+        memcpy(*p_testperm, v.testperm, NT * sizeof(int));
+
+    free(v.testperm);
```
[the remark this answers is not in the document]
Comment (Kanma) [reply solver/verify.c:1599]: To add to my comment:

The only invocation of ```tweak2()``` I found where ```testperm``` isn't NULL is in ```solver_tweak2()``` which immediately free the memory without using it, with a "Nobody uses it, right?" comment.
